spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BCECriterion.scala

```diff
@@ -87,27 +131,32 @@ class BCECriterion[@specialized(Float, Double) T: ClassTag]
     require(input.nElement() == target.nElement(),
       "input and target should have the same dims." +
         s"input dim(${input.nElement()})" +
-        s"taget dim(${target.nElement()})")
+        s"target dim(${target.nElement()})")
 
-    if (null != weights && target.dim() != 1) {
-      weights = weights.view(1, target.size(2)).expandAs(target)
-    }
-
-    val norm = if (sizeAverage) 1.0 / input.nElement() else 1.0
+    val nElement = input.nElement()
+    val norm = if (sizeAverage) 1.0 / nElement else 1.0
 
     gradInput.resizeAs(input)
 
-    val func = new TensorFunc6[T] {
-      override def apply(data1: Array[T], offset1: Int, data2: Array[T], offset2: Int,
-                         data3: Array[T], offset3: Int): Unit = {
-        val x = ev.toType[Double](data2(offset2))
-        val y = ev.toType[Double](data3(offset3))
-        data1(offset1) = ev.fromType(-norm * (y - x) / ((1.0 - x + eps) * (x + eps)))
+    if (nElement < 200 || (null != weights && weights.nElement() < 200)) {
+      val func = new TensorFunc6[T] {
+        override def apply(data1: Array[T], offset1: Int, data2: Array[T], offset2: Int,
+                           data3: Array[T], offset3: Int): Unit = {
+          val x = ev.toType[Double](data2(offset2))
+          val y = ev.toType[Double](data3(offset3))
+          data1(offset1) = ev.fromType(-norm * (y - x) / ((1.0 - x + eps) * (x + eps)))
+        }
       }
+      DenseTensorApply.apply3(gradInput, input, target, func)
+    } else {
+      // - (1 - x + eps)*(x + eps) = x^2 - x - eps - eps^2
+      // eps^12 is negligible
+      buffer.copy(input).square().sub(input).sub(ev.fromType(eps))
```

Comment:
use `buffer.pow(input, 2)` instead or add an `square(x: Tensor[T])` interface to Tensor.